### ISSUE TITLE
Refactoring iterative methods with an update!() function

### ISSUE THREAD

**Issue by pkofod** [JuliaOpt/Optim.jl#254]
There was a discussion over at https://gitter.im/JuliaML/chat about how to train models in the JuliaML project. It was mentioned that Optim might be useful. However, when training, we often do not "run until convergence" in the sense we do it here. We might want to calculate different things in between, and we might not want to end it when the gradient is below some threshold. The idea that was discussed was to refactor Optim solvers such that we get something along the lines of 

``` julia
function optimize(...)
    s = initialize_state(mo, o)
    while !converged && iteration < o.iteration
        update!(s, ...)
        ... = converged(....)
    end
    return MultivariateOptimizationResults(...)
end
```

and then each solver has to implement `initialize_state` and `update!` instead of implementing an `optimize(df, initial_x, mo, o)` method. We could also add a callback after update to allow for more flexible stopping.

Any thoughts on this? I've had the request before from other people.

**Comment by johnmyleswhite** [comment]
I've done a lot of this kind of work with rigged up local copies of Optim. I think you need to do this and also provide a mechanism for providing all of the intermediate data structures, so you don't waste time allocating them on each iteration. Definitely worth doing.

**Comment by KristofferC** [comment]
Sounds a bit like https://github.com/spencerlyon2/IterationManagers.jl

**Comment by mlubin** [comment]
This is generally useful. Would have erased the need for this [juliacon talk](https://www.youtube.com/watch?v=qAD0ZXriBJs).

**Comment by pkofod** [comment]
Thanks for the quick feedback/comments. I think it's clear that this is happening. I hope it will clean up some code, as well as increase the usefulness of the package!

> I've done a lot of this kind of work with rigged up local copies of Optim. I think you need to do this and also provide a mechanism for providing all of the intermediate data structures, so you don't waste time allocating them on each iteration. Definitely worth doing.

Great; it will be next on my list, but for v0.7. I'll cook up a PR, but I'll be very happy to take ideas and criticism from people who've used this kind of functionality.

> Sounds a bit like https://github.com/spencerlyon2/IterationManagers.jl

Yes, it's kind of related to that. 

> This is generally useful. Would have erased the need for this juliacon talk.
I didn't see that talk all the way through, maybe I should before I start working on this :)

**Comment by cortner** [comment]
Kind of a side issue, but something worth thinking about? 

I have several objects in my own codes that use `update!` and I suspect many others have as well. It is just a very canonical term. E.g., concerning preconditioners, I would have liked to replace `precondprep!` with `update!` but didn't because that would require a generic interface. I wonder therefore (but have no clue whether it is feasible) whether an  `update!` function (maybe empty) could go into `Base`, so that it can be imported.

**Comment by KristofferC** [comment]
Or just don't export it.

**Comment by Evizero** [comment]
@cortner I disagree with that a bit (for biased reasons). A function defined in Base comes with the implicit contract to be used in a particular way and mean a particular thing. At least that is how I understand it. That would then probably not be what we have in mind here.

It seems that the JuliaOpt community is on board with this idea of introducing `update!`, and we at JuliaML would welcome this change and follow their lead accordingly (we need Optim, but Optim doesn't need us). That would mean a great deal for our effort to streamline things a bit and reduce/avoid redundancy.

**Comment by pkofod** [comment]
> Or just don't export it.

Exactly, I didn't intend to export it. If people want to build custom projects using an update, they'll have to qualify it (`Optim.update!()`). The alternative is to choose another name; `update_x!` is a possibility if it needs to be "special".

**Comment by cortner** [comment]
I don't have a problem with `update!` and even if I did I wouldn't really want to impose my view on how `Optim` is developed.  I appreciate that I can always write `Optim.update!` to specify which `update!` I mean, so possibly what I had in mind will case very limited problems and more related to other uses of `update!` than this specific one. I disagree with `update_x!` though; for me the point about dispatch is that the specialisation comes from the arguments.

**Comment by ahwillia** [comment]
@pkofod - do you have a sense of a timeline for implementing this? I would really like to use this (even on an experimental branch) sooner rather than later. I would be more than happy to contribute to this effort if you think I could help, but you have a better sense of the code than me -- I don't want to slow you down.

**Comment by pkofod** [comment]
I would say it's next on my list. So.. Beginning of next week?

**Comment by tbreloff** [comment]
Woohoo! Looking forward to using this

**Comment by pkofod** [comment]
I'll push when I have more, but a version of this works locally for Gradient Descent. I'll work a bit more on it, and then push for feedback.

**Comment by tbreloff** [comment]
I'm thinking through this problem right now as I want to use it with Reinforce.jl.  I have the Cross Entropy Method implemented mainly for testing purposes, and I'd like to attempt to make this a Optim.Optimizer, but I could use some guidance on exactly what needs to be done.

Here's my current version:  https://github.com/tbreloff/Reinforce.jl/blob/master/test/cem_gym.jl#L87

I tried to clearly mark the sections that I thought could be split out into a generalized solver loop: init, update, trace, convergence.  Does this line up with what you're thinking for your refactor @pkofod?  If I was to make this an Optimizer, what changes would I need to make?

**Comment by pkofod** [comment]
Looking at it quickly, I don't think you should have to change too much.

Basically, right now a gradient free optimizer needs a
- `method_string(method::Method) = "Nicely Formatted Method"`, this is for the output
- a `MethodState` that contains all the variables you don't want to have to allocate memory for each time AND the fields required for `MultivariateOptimizationResult`.
- an`initialize_state(method::Method, options, f::Function, initial_x)`
- an `initialize_state(method::Method, options, d, initial_x::Array) = initialize_state(method, options, d.f, initial_x)` (this is due to the whole `DifferentiableFunction`-machinery)
- an `update!{T}(f::Function, state::ParticleSwarmState{T}, method::ParticleSwarm) where `T` is `eltype(initial_x)`
- an `update!(d, state::ParticleSwarmState, method::ParticleSwarm) = update!(d.f, state, method)` (again to handle the `DifferentiableFunction`-thing)
- an `assess_convergence(state, options)`
- a `trace!(tr, state, iteration, method, options)`, and 
- `after_while!(d, state, method, options)` that basically just does post optimization stuff like saving your animation

A (very much non-working, as I'm not sure what everything is) sketch can be seen here https://gist.github.com/pkofod/41ef3bceab07586e62a2cd01590bc56e . So I think maybe a lot of the `solver` fields would be fed into the `State`. I can try to make a working `CrossEntropy` optimizer based on your code, but I don't have time at this exact moment.

I should write a better description of this in the developer docs in the PR :)

**Comment by tbreloff** [comment]
Thanks @pkofod I'll tackle this today and report back.

**Comment by pkofod** [comment]
Great. Nothing is set in stone, so feedback is very welcome.
